### PR TITLE
refactor: Move actor state into propagator state

### PR DESCRIPTION
**Pull request by andiwand**
I propose to move the actor state into the propagator state. This allows us to copy the whole propagator state and to start the propagation from a specific state after it stopped before.

Intention of this change is that algorithms may want to steer the propagation from the outside rather than from within by using an actor.

### CODE REVIEW

**Comment by andiwand** [comment]
@paulgessinger @benjaminhuth @asalzburger do you have some early feedback on this?

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 1]
```diff
@@ -324,10 +320,15 @@ class CombinatorialKalmanFilter {
...
-    // The source link container type
     /// Broadcast the result_type
     using result_type = CombinatorialKalmanFilterResult<traj_t>;
 
+    /// Fitted states that the actor has handled.
+    traj_t* fittedStates{nullptr};
+
+    /// This is used internally to store candidate trackstates
+    std::shared_ptr<traj_t> stateBuffer;
+
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:331]: right I changed this very early on while refactoring and this ended up not being necessary after all

the "result" (actor state) lives now within the propagator state and I had troubles to set state variables from the outside which is necessary in this case. sometimes we do have state in the actor but I guess just immutable? might be good to get rid of these all the way and to put it into the "result".

I surrounded this issue now by having a `makeState` method in the `Propagator` which lets you create a state, then mutate it and then call the propagator on this state. I think that covers our use cases.

having a `propagate` method which can be called multiple times also lets us steer the `Propagator` from the outside which will come handy for track finding in the future

[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:331]: I am still not too happy with the name "result" but I kept it for now as it is still returned after the propagation for the usual convenience functions

[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:346]: So the thinking was if it's immutable it can be a member, if it's mutable it's *state*. We didn't originally foresee actor-specific state, but started abusing the result for this early on.

My thinking was that we should eventually (tm) just rename *result* into *state* or *actorState* or whatever. This is orthogonal to putting said state into the main propagator state or not.

File: Core/include/Acts/Propagator/Propagator.hpp [hunk 2]
```diff
@@ -320,6 +353,15 @@ class Propagator final {
...
 
  public:
+  /// @brief Short-hand type definition for propagation state derived from
+  ///        an action list
+  ///
+  /// @tparam action_list_t List of propagation action types
+  ///
+  template <typename propagator_options_t, typename action_list_t>
+  using action_list_t_state_t =
+      typename state_type_helper<propagator_options_t, action_list_t>::type;
```
Comment (asalzburger) [inline Core/include/Acts/Propagator/Propagator.hpp:363]: Where do you need this forward?

File: Core/include/Acts/Propagator/Propagator.hpp [hunk 3]
```diff
@@ -437,41 +426,58 @@ class Propagator final {
...
   template <typename parameters_t, typename propagator_options_t,
             typename target_aborter_t = SurfaceReached,
             typename path_aborter_t = PathLimitReached>
+  auto makeState(const parameters_t& start, const Surface& target,
+                 const propagator_options_t& options) const;
+
+  template <typename propagator_state_t, typename propagator_options_t>
   Result<
-      action_list_t_result_t<StepperBoundTrackParameters,
+      action_list_t_result_t<StepperCurvilinearTrackParameters,
```
Comment (asalzburger) [inline Core/include/Acts/Propagator/Propagator.hpp:459]: Can you explain, why this change of type is needed here?

File: Core/include/Acts/Propagator/Propagator.ipp [hunk 4]
```diff
@@ -26,23 +25,22 @@ auto Acts::Propagator<S, N>::propagate_impl(propagator_state_t& state,
   // Navigator initialize state call
   m_navigator.initialize(state, m_stepper);
   // Pre-Stepping call to the action list
-  state.options.actionList(state, m_stepper, m_navigator, result, logger());
```
Comment (asalzburger) [inline Core/include/Acts/Propagator/Propagator.ipp:29]: I like this - removes this awkward circular stuffing back of the result to the one that would create the result.

File: Core/include/Acts/Propagator/Propagator.ipp [hunk 5]
```diff
@@ -26,23 +25,22 @@ auto Acts::Propagator<S, N>::propagate_impl(propagator_state_t& state,
...
-                               logger())) {
+  if (!state.options.abortList(state, m_stepper, m_navigator, logger())) {
     // Stepping loop
     ACTS_VERBOSE("Starting stepping loop.");
 
     terminatedNormally = false;  // priming error condition
 
     // Propagation loop : stepping
-    for (; result.steps < state.options.maxSteps; ++result.steps) {
+    for (; state.steps < state.options.maxSteps; ++state.steps) {
```
Comment (asalzburger) [inline Core/include/Acts/Propagator/Propagator.ipp:43]: Also, I like that - it becomes a state property.

File: Core/include/Acts/Propagator/Propagator.ipp [hunk 6]
```diff
@@ -106,37 +103,47 @@ auto Acts::Propagator<S, N>::propagate(const parameters_t& start,
...
-    const parameters_t& start, const propagator_options_t& options,
-    bool makeCurvilinear,
-    action_list_t_result_t<StepperCurvilinearTrackParameters,
-                           typename propagator_options_t::action_list_type>&&
-        inputResult) const
+    const parameters_t& start, const Surface& target,
+    const propagator_options_t& options) const
     -> Result<action_list_t_result_t<
-        StepperCurvilinearTrackParameters,
+        StepperBoundTrackParameters,
```
Comment (asalzburger) [inline Core/include/Acts/Propagator/Propagator.ipp:126]: Is this PR mixing two developments? Or why is the change here happening?

File: Core/include/Acts/Propagator/Propagator.ipp [hunk 7]
```diff
@@ -154,7 +161,9 @@ auto Acts::Propagator<S, N>::propagate(
   auto eOptions = options.extend(abortList);
   using OptionsType = decltype(eOptions);
   // Initialize the internal propagator state
-  using StateType = State<OptionsType>;
+  using StateType =
+      action_list_t_state_t<OptionsType,
+                            typename propagator_options_t::action_list_type>;
```
Comment (asalzburger) [inline Core/include/Acts/Propagator/Propagator.ipp:166]: Got it, answers the question from above.

File: Core/include/Acts/TrackFitting/GaussianSumFitter.hpp [hunk 8]
```diff
@@ -339,7 +341,6 @@ struct GaussianSumFitter {
       actor.m_cfg.inputMeasurements = &inputMeasurements;
       actor.m_cfg.inReversePass = true;
       actor.m_cfg.logger = m_actorLogger.get();
-      actor.setOptions(options);
```
Comment (asalzburger) [inline Core/include/Acts/TrackFitting/GaussianSumFitter.hpp:342]: I missed that, why do the options disappear here?

File: Core/include/Acts/TrackFitting/GaussianSumFitter.hpp [hunk 9]
```diff
@@ -349,47 +350,33 @@ struct GaussianSumFitter {
...
+                                          MultiStepperSurfaceReached>(
+              params, target, bwdPropOptions);
 
       assert(
           (fwdGsfResult.lastMeasurementTip != MultiTrajectoryTraits::kInvalid &&
            "tip is invalid"));
 
-      auto proxy =
-          r.fittedStates->getTrackState(fwdGsfResult.lastMeasurementTip);
+      auto proxy = trackContainer.trackStateContainer().getTrackState(
```
Comment (asalzburger) [inline Core/include/Acts/TrackFitting/GaussianSumFitter.hpp:364]: Also this seems not directly related to the PR?

**Review by asalzburger — CHANGES_REQUESTED** [review]
I like the PR a lot, particularly it moves us closer to eventually combine actors/aborters at some point?

Anyhow, I do not fully get the changes concerning TrackContainer and Bound/Curvilinear type change - are those knock-on effects?

File: Core/include/Acts/Propagator/Propagator.ipp [hunk 10]
```diff
@@ -106,37 +103,48 @@ auto Acts::Propagator<S, N>::propagate(const parameters_t& start,
...
         typename propagator_options_t::action_list_type>> {
-  // Type of track parameters produced by the propagation
-  using ReturnParameterType = StepperCurvilinearTrackParameters;
+  static_assert(
+      std::is_copy_constructible<StepperCurvilinearTrackParameters>::value,
+      "return track parameter type must be copy-constructible");
 
-  static_assert(std::is_copy_constructible<ReturnParameterType>::value,
-                "return track parameter type must be copy-constructible");
+  auto state = makeState<parameters_t, propagator_options_t, path_aborter_t>(
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/Propagator.ipp:110]: Do you need the explicit template parameters or are they deduced?

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 11]
```diff
@@ -1461,23 +1459,17 @@ class CombinatorialKalmanFilter {
...
-    r.stateBuffer = stateBuffer;
-    r.stateBuffer->clear();
+    r.stateBuffer = std::make_shared<traj_t>();
+
+    auto propagationResult = m_propagator.template propagate(propState);
 
-    auto result = m_propagator.template propagate<
-        start_parameters_t, decltype(propOptions), PathLimitReached>(
-        initialParameters, propOptions, false, std::move(inputResult));
+    auto result = m_propagator.template makeResult(
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1471]: I like this pattern much better than what we had before.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 12]
```diff
@@ -1461,23 +1459,17 @@ class CombinatorialKalmanFilter {
...
+    auto propState =
+        m_propagator.template makeState(initialParameters, propOptions);
 
+    auto& r = propState.template get<CombinatorialKalmanFilterResult<traj_t>>();
     r.fittedStates = &trackContainer.trackStateContainer();
-    r.stateBuffer = stateBuffer;
-    r.stateBuffer->clear();
+    r.stateBuffer = std::make_shared<traj_t>();
+
+    auto propagationResult = m_propagator.template propagate(propState);
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1469]: Does this need the `template` if there are no explicit template parameters?

File: Core/include/Acts/TrackFitting/GaussianSumFitter.hpp [hunk 13]
```diff
@@ -276,29 +274,33 @@ struct GaussianSumFitter {
...
-        return m_propagator.propagate(params, fwdPropOptions, false,
-                                      std::move(inputResult));
       } else {
-        return m_propagator.propagate(sParameters, fwdPropOptions, false,
-                                      std::move(inputResult));
+        params = sParameters;
       }
+
+      auto state =
+          m_propagator.template makeState<MultiComponentBoundTrackParameters,
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/GaussianSumFitter.hpp:292]: Explicit template parameters needed?

File: Core/include/Acts/TrackFitting/GaussianSumFitter.hpp [hunk 14]
```diff
@@ -276,29 +274,33 @@ struct GaussianSumFitter {
...
+                                          decltype(fwdPropOptions)>(
+              *params, fwdPropOptions);
+
+      auto& r = state.template get<typename GsfActor::result_type>();
+      r.fittedStates = &trackContainer.trackStateContainer();
+
+      auto propagationResult = m_propagator.propagate(state);
+
+      return m_propagator
+          .template makeResult<decltype(state), decltype(fwdPropOptions)>(
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/GaussianSumFitter.hpp:302]: Also this should be deduced, no?

File: Core/include/Acts/TrackFitting/GaussianSumFitter.hpp [hunk 15]
```diff
@@ -349,47 +350,35 @@ struct GaussianSumFitter {
...
-              std::declval<decltype(inputResult)>()));
-
-      auto& r = inputResult.template get<typename GsfActor::result_type>();
-
-      r.fittedStates = &trackContainer.trackStateContainer();
+      const auto& params = *fwdGsfResult.lastMeasurementState;
+      auto state =
+          m_propagator.template makeState<MultiComponentBoundTrackParameters,
+                                          decltype(bwdPropOptions),
+                                          MultiStepperSurfaceReached>(
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/GaussianSumFitter.hpp:357]: Same here?

File: Core/include/Acts/TrackFitting/GaussianSumFitter.hpp [hunk 16]
```diff
@@ -349,47 +350,35 @@ struct GaussianSumFitter {
...
 
-      const auto& params = *fwdGsfResult.lastMeasurementState;
+      auto propagationResult = m_propagator.propagate(state);
 
-      return m_propagator.template propagate<std::decay_t<decltype(params)>,
-                                             decltype(bwdPropOptions),
-                                             MultiStepperSurfaceReached>(
-          params, target, bwdPropOptions, std::move(inputResult));
+      return m_propagator
+          .template makeResult<decltype(state), decltype(bwdPropOptions)>(
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/GaussianSumFitter.hpp:380]: And here

File: Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp [hunk 17]
```diff
@@ -241,11 +241,11 @@ struct Gx2FitterResult {
 /// @tparam measDim Number of dimensions of the measurement
 /// @tparam traj_t The trajectory type
 ///
-/// @param trackStateProxy is the current track state
+/// @param trackStateProxy is the track state proxy
 /// @param result is the mutable result/cache object
 /// @param logger a logger instance
 template <std::size_t measDim, typename traj_t>
-void collector(typename traj_t::TrackStateProxy& trackStateProxy,
+void collector(const typename traj_t::TrackStateProxy& trackStateProxy,
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp:248]: Could this be a `ConstTrackStateProxy`?

File: Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp [hunk 18]
```diff
@@ -643,19 +643,21 @@ class Gx2Fitter {
       gx2fActor.actorLogger = m_actorLogger.get();
       gx2fActor.nUpdate = nUpdate;
 
-      typename propagator_t::template action_list_t_result_t<
-          CurvilinearTrackParameters, Actors>
-          inputResult;
+      auto propagatorState =
+          m_propagator.template makeState(params, propagatorOptions);
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp:647]: `template` needed?

**Review by paulgessinger — COMMENTED** [review]
There are a number of cases where I wonder if you need to supply the explicit template parameters of you could let the compiler figure them out. Did you try this out? 

Other than that, do you need the `template` if you don't supply explicit arguments? I thought this is just needed to disambiguate what `<>` means syntactically.

File: Core/include/Acts/TrackFitting/GaussianSumFitter.hpp (hunk 15 above)
Comment (andiwand) [reply Core/include/Acts/TrackFitting/GaussianSumFitter.hpp:357]: I think here it is not possible because we need to specialize on `MultiStepperSurfaceReached`

**Comment by andiwand** [comment]
> There are a number of cases where I wonder if you need to supply the explicit template parameters of you could let the compiler figure them out. Did you try this out?
> 
> Other than that, do you need the `template` if you don't supply explicit arguments? I thought this is just needed to disambiguate what `<>` means syntactically.

big thanks for your comments @paulgessinger ! I am cleaning this out

**Review by asalzburger — APPROVED** [review]
I am happy with it now.

**Review by paulgessinger — COMMENTED** [review]
Found one more case, but not critical

**Review by paulgessinger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [6f8e9b85f954b8d179deff3d9db7e0e5fcf6db50]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35380853) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6754520)!
